### ISSUE TITLE
Combine info on changes Conventions

### ISSUE THREAD

**Issue by erget** [cf-convention/cf-conventions#369]
Some discussion with @JonathanGregory on minting 1.10 brought us to this issue. I've volunteered to look at this at the upcoming workshop and won't fill this with too much details, but here's some context so I'll remember what it's about:

> Do you think that `rules.md` is the right place to document these? It sounds to me more like modalities of working with GitHub, in which case `CONTRIBUTING.md` might be a better home for these guidelines.

> ...These two documents [`rules.md` and `CONTRIBUTING.md`] overlap somewhat in purpose. Could we merge them? In order to do that, we would have to put the [rules for making enhancements](http://cfconventions.org/rules.html) and [rules for correcting defects](http://cfconventions.org/errors.html) into the same file, along with the contents of [`CONTRIBUTING`](https://github.com/cf-convention/cf-conventions/blob/main/CONTRIBUTING.md).

**Comment by JonathanGregory** [comment]
## Background

Daniel @erget began this issue two years ago. I'm reviving it now in an attempt to bring it to a conclusion, following discussions in the last two CF committee meetings. As Daniel wrote above, he and I had discussed merging [`CONTRIBUTING`](https://github.com/cf-convention/cf-conventions/blob/main/CONTRIBUTING.md) in this (conventions) repo and [`rules`](https://cfconventions.org/rules.html) in the website repo, because both of them are about how to change the conventions. Since moderation of proposals and use of labels are both aspects of the process to be documented, the discussion in the open issues #151 on moderation of proposals and #440 on labels will inform this issue as well.

I believe that this issue should supersede [website issue #102](https://github.com/cf-convention/cf-convention.github.io/issues/102) on "refactoring the rules". Issue 102 is in the website repo because the rules for changing the conventions were originally part of the governance arrangements. These days, the conventions are only one of three things which we change (conventions, vocabulary and website/governance), each with its own repo. Therefore I think it's logical to discuss those rules in _this_ repo and issue. Daniel drafted some changes in his [website PR #103](https://github.com/cf-convention/cf-convention.github.io/pull/103). That PR was merged by accident, which Daniel undid with some git enchantment, so those changes have _not_ been implemented. I propose that we should continue with the discussion of Daniel's draft text in this issue, as well as the subsequent additions in [website issue #102](https://github.com/cf-convention/cf-convention.github.io/issues/102)  that aren't in his PR [website PR #103](https://github.com/cf-convention/cf-convention.github.io/pull/103).

Website issue #102 includes a diagram of the process as well. For simplicity, I propose that we should _not_ talk about the diagram in this issue, because the later [website issue #217](https://github.com/cf-convention/cf-convention.github.io/issues/217) is a good place to do that.

As you see, part of my aim with this issue is to tidy up and provide an overview of this set of unfinished issues. To summarise:

#369 (this issue). Merge `CONTRIBUTING` and `Rules`. If we agree to do that, we can continue in this issue (or start a new one), on the text of that unified document. The text will be informed by the following five items:

[website issue #102](https://github.com/cf-convention/cf-convention.github.io/issues/102) and [associated PR](https://github.com/cf-convention/cf-convention.github.io/pull/103) on clarifying the rules.

#151 on moderation of proposals for conventions changes.

#328 on the rules for deprecating content

#440 ~on labels in the conventions repository~ _has been concluded_

[Discussion #349](https://github.com/orgs/cf-convention/discussions/349), about a couple of aspects of the change process (cooling-off period and need for test file)

#348 has recently been reinvigorated, concerning provision of test files, and is therefore relevant to the rules.

#325 ~on updating the checklist in the conventions PR template~ _has been concluded_.

[website issue #217](https://github.com/cf-convention/cf-convention.github.io/issues/217) on a diagram of the change process.

## Proposal

I propose that we

* Delete the links to [rules for making changes to the CF Conventions](https://cfconventions.org/rules.html) and [rules for correcting errors in the CF documents](https://cfconventions.org/errors.html) from the [governance page](https://cfconventions.org/governance.html), leaving just the links to the CF constitution and the 2006 white paper which preceded it.

* Delete `rules.html` and `errors.html` from the website repo.

* Revise the `CONTRIBUTING` document by combining the two sets of rules, including the short revision history of the rules which currently appears on the governance page, with the current text of `CONTRIBUTING`.

* Make a link to the new document on the [conventions page](https://cfconventions.org/conventions.html). This would be parallel to the link to the instructions for proposing changes to vocabulary which appears on the [vocabularies page](https://cfconventions.org/vocabularies.html).

* Make links to both the new `CONTRIBUTING` and the vocabulary rules on the [discussion page](https://cfconventions.org/discussion.html), and as quick links from the [CF home page](https://cfconventions.org).

This means that, whether you look at the website or the repo, you will always be directed to the same `CONTRIBUTING` file for instructions on proposing changes to the conventions, which will be in this (conventions) repo, the relevant one. (Among others, @mwengren has argued for arrangements like this.)

What are your opinions?

**Comment by larsbarring** [comment]
Excellent, I agree that this is a good way forward. Thanks for initiating this! In particular I like unifying the rules in the web pages and the github repo**s**.

**Comment by JonathanGregory** [comment]
In [website issue 545](https://github.com/cf-convention/cf-convention.github.io/issues/545#issuecomment-2376767082) I have posted a recipe for making a PR, assuming no experience with `git` or GitHub. The same procedure applies to this (conventions) repo. Following a comment of @DocOtak, I suggest that it would be useful to put those instructions in a webpage, and refer to it from the `CONTRIBUTING.md` files of both the conventions and website repos.

**Comment by JonathanGregory** [comment]
Dear all

As proposed 1.5 years ago in this issue, I have made a [draft](https://github.com/cf-convention/cf-conventions/wiki/Draft-consolidated-rules-and-CONTRIBUTING-file) of a consolidated document by merging the existing rules for enhancements and defects and the contents of `CONTRIBUTING`, with some new text, elimination of duplication, rearrangement and editing. The [proposal (as above)](https://github.com/cf-convention/cf-conventions/issues/369#issuecomment-2315019761) is that this single document, when agreed, will replace the existing documents.

Please read the [draft](https://github.com/cf-convention/cf-conventions/wiki/Draft-consolidated-rules-and-CONTRIBUTING-file) and comment. If your comment is about matters of content, please comment here in the issue. Comments about wording can be made in the draft, which is a wiki page. There's no automatic way to identify who made changes, so please identify your edits somehow.

Some points to note:

* The rules are not clear about whether the author of a PR should be allowed to merge it (once the change is agreed). I think it's preferable for someone else to do it, just for the sake of another quick look at it, but I don't know if it should be required. Also, maybe it's different for `enhancements` and `defects`. What do you think?

* The material about authorship, dormant issues and test files is new, and follows previous discussions and agreements by the committee.

* Rules are included for our recent practice of devolving really difficult issues to a group of interested people to sort out through meetings.

* Moderators are no longer expected to provide summaries of the content of the discussion, since this is rather demanding and puts people off from volunteering to moderate. The main purpose of a moderator is to keep the discussion moving towards a conclusion. It's not always necessary to have a moderator.

* The procedure for dealing with flaws in the convention has been revised following experience and discussion. It now envisages flaws which have existed for a long time (not resulting from changes in the recent version), and clarifies that a remedy for a flaw should include a requirement that no more data be written according to the flawed convention.

* The proposed infrastructure guide should be linked from this rules document, once it's been compiled.

* We still need to clarify the rules for changing to GitHub procedures in this repo, and for changing the rules themselves. I propose that we do that subsequently in a new issue.

* [website issue #217](https://github.com/cf-convention/cf-convention.github.io/issues/217) proposes a diagram of the change process. This issue is dormant and could be revived once we've agreed the changes to the rules.

I hope that, if we can agree this consolidated document with these contents, we will be able to close the following as well as this issue:

[website issue #102](https://github.com/cf-convention/cf-convention.github.io/issues/102) and [associated PR](https://github.com/cf-convention/cf-convention.github.io/pull/103) on clarifying the rules.

#151 on moderation of proposals for conventions changes.

#328 on the rules for deprecating content.

[Discussion #349](https://github.com/orgs/cf-convention/discussions/349), about a couple of aspects of the change process (cooling-off period and requirement for test file).

Best wishes

Jonathan

**Comment by ChrisBarker-NOAA** [comment]
> The rules are not clear about whether the author of a PR should be allowed to merge it (once the change is agreed). I think it's preferable for someone else to do it, just for the sake of another quick look at it, but I don't know if it should be required. Also, maybe it's different for enhancements and defects. What do you think?

With a volunteer based organization like this, I think minimal friction is key the actually getting things done. [case in point -- this issue is almost 4 yrs old!] And git is a fabulous system for keeping track of changes, who made them, and being able to role back things if need be.

With that philosophy in mind,  the author of a PR should be allowed to merge it. 

I agree that it's better that it gets a last look by someone else, but that shouldn't hold up progress! 

As for the draft proposal:

> Detailed review of the text textual changes may take place either in the issue or in a pull request

I think the PR is the right place to hash out the details of the text -- it's much easier to review in context and comment on specific wording. 

However, it's a fuzzy line between text and concepts, and:

" the scientific conceptual development of the proposal is easily found in one place, i.e. the GitHub issue."

Is a good point. I'm not sure how to make that more clear, but maybe:

Once there is a decision on the concepts, i.e. the *what* to do is decided, the final textual changes should be  proposed and reviewed via a pull request. Comments in pull requests should be limited to copy editing, and not be used for discussion of the ideas of issue. If during text review conceptual issues arise, they should be discussed in the Issue, not the PR. This is so that the conceptual development of the proposal is easily found in one place, i.e. the GitHub issue.

I've only given the rest of the draft a very quick read, but it's looking good to me! 

Thanks!

**Comment by JonathanGregory** [comment]
Dear @ChrisBarker-NOAA 

Thanks for your helpful comments. I've made some changes to the [draft](https://github.com/cf-convention/cf-conventions/wiki/Draft-consolidated-rules-and-CONTRIBUTING-file).

Regarding who can merge PRs, I have put "the moderator or anyone with the necessary GitHub permissions", and added, "It may be useful for someone other than the author of the pull request to check and merge it, to reduce the possibility of oversights, but this is not a requirement."

Regarding where text should be discussed, I've put:

A pull request may be opened once the discussion has reached the stage of discussing the text in detail, and review of the textual changes may take place either in the pull request or in the issue, whichever is clearer.  For instance, it can be clearer to comment on specific wording in the context where it occurs in the pull request, but clearer to make suggestions in the issue for global changes to wording or other changes which affect several parts of the text at once.  Comments in pull requests should not be used for discussion of the issue in general.  This is so that the conceptual development of the proposal is easily found in one place, i.e. the GitHub issue.

What do you think?

Also, I've inserted a contents section at the start.

Jonathan

**Comment by erget** [comment]
Working my way thru todos in order to close this neatly.

From https://github.com/cf-convention/cf-conventions/issues/369#issuecomment-2315019761:

- [x] Close https://github.com/cf-convention/cf-convention.github.io/issues/102

Will need to do:
- [x] Review https://github.com/cf-convention/cf-conventions/wiki/Draft-consolidated-rules-and-CONTRIBUTING-file, esp. in light of https://github.com/cf-convention/cf-conventions/issues/369#issuecomment-4049293460
- [x] Convert it into PR
- [x] Check cross-references and deletions, this may spawn PRs for other repos that need to be merged at same time

So we don't forget here are things we'll need to consolidate the changes here later: [_I have closed the following three issues as concluded today, along with the present issue._ JMG 5.6.26]

- https://github.com/cf-convention/cf-conventions/issues/151 on moderation of proposals for conventions changes.
- https://github.com/cf-convention/cf-conventions/issues/328 on the rules for deprecating content.
- https://github.com/orgs/cf-convention/discussions/349, about a couple of aspects of the change process (cooling-off period and requirement for test file).

Will come back to this tomorrow.

**Comment by erget** [comment]
https://github.com/cf-convention/cf-conventions/wiki/Draft-consolidated-rules-and-CONTRIBUTING-file looks good for me; I've made some minor improvements. I had also promised @JonathanGregory to have a think about adding a paragraph or 2 about the fact that CF depends on voluntary effort and that if a large volume of requests come in, we can't be expected to process them all in any given time frame.

I believe that the process is clearly very manual in nature, as we depend on human judgment for evaluating every change we make to the Conventions, and the description of how changes work is clear and verbose. Therefore I won't add anything there. There might be cause to add something to the code of conduct if we feel that there's potential that contributors could become unsatisfied and start demanding faster action (as an impatient person, I can relate to that!). Given the voluntary nature of our contributions, I don't think that would be appropriate and I think that this is well-reflected in [the opening to the code of conduct](https://github.com/cf-convention/cf-conventions?tab=coc-ov-file#readme) already, although we could sharpen it if we notice that this is occuring:
> As contributors and maintainers of this project, we pledge to respect all people who contribute through reporting issues, posting feature requests, updating documentation, submitting pull requests or patches, and other activities.

On other notes - it looks like https://github.com/cf-convention/cf-conventions/issues/369#issuecomment-4049293460 is well-addressed in the curent draft. ☑️ 

I won't manage it any more this week but I think this can now be converted into a PR (possibly multiple as I think we might be crossing repositories now) so that we can merge the new text and publish across the board.

**Comment by JonathanGregory** [comment]
Thanks for your review and improvements, Daniel @erget. I have made a PR #645 to replace the existing contents of `CONTRIBUTING.md` with the new merged contents including the rules, and also making small amendments at the start of `README.md`, which refers to the rules.

Everyone is invited to consider and comment on the new drafts of [`CONTRIBUTING.md`](https://github.com/JonathanGregory/cf-conventions/blob/issue369/CONTRIBUTING.md) and [`README.md`](https://github.com/JonathanGregory/cf-conventions/blob/issue369/README.md).

Antonio @cofinoa, would it be possible for the [GitHub Pages site for this repo](https://cfconventions.org/cf-conventions) to include rendered `html` of `CONTRIBUTING` and `README`? These would be nicer to link from the website than the `md` files from the blob, I think.

I have opened a related [issue in the website repo](https://github.com/cf-convention/cf-convention.github.io/issues/632) about linking the new merged file.

**Comment by davidhassell** [comment]
Very nice, Jonathan!

I approve of the PR in its current form.

I have made a couple of trivial suggestions on the PR #645 (which you are free to reject).

The only other question I had was in this text:

> After that period, the issue should be closed by the manager of the CF conventions, the manager of CF standard names or any member of the CF conventions committee, who will make the change and/or merge the pull request.

I'm not sure what the "manager" role is. Again, ignore if this is my ignorance :)

**Comment by JonathanGregory** [comment]
Thanks for your review, @davidhassell. I've accepted your changes, which are improvement. The "manager" was a mistake for "secretary of the CF conventions" i.e. you! I've changed it, and also "secretary of CF standard names" viz Alison.

There are no outstanding concerns on these changes to the rules. If anyone else has time to check that the text is clear, correct and complete, please do so. These changes will be accepted in three weeks (on 5th June) if no new concerns are raised before then. Thanks.

**Comment by ethanrd** [comment]
Thanks Jonathan @JonathanGregory! And thanks to everyone that has contributed. An impressive amount of work clearly went into this thorough updating of the rules.

I also approve the PR as it stands.

I made two minor edit suggestions in the PR. I also have a slightly more substantive suggestion for the CONTRIBUTING document that I'm happy to drop or postpone to another issue if it might be substantive enough to delay the conclusion of this issue/PR. Here it is:

In last paragraph of the introduction (before Contents), it might be useful to mention that contributors should not feel they need to read the entire document before starting to contribute. Reading the introductory paragraphs of the "Rules for Changes to the CF Conventions" should (perhaps?) be enough to get started. Though if we add that, it might be good to mention the 3-week rule in those introductory paragraphs.

**Comment by ethanrd** [comment]
@JonathanGregory - In regards to your question for Antonio about html renderings of these documents. The README file is presented by default at the bottom of the repo's main page and the CONTRIBUTING file is presented in another tab "beside" the README. They can each be reached directly here:
- https://github.com/cf-convention/cf-conventions?tab=readme-ov-file
- https://github.com/cf-convention/cf-conventions?tab=contributing-ov-file

The CODE_OF_CONDUCT file is presented similarly:
- https://github.com/cf-convention/cf-conventions?tab=coc-ov-file

**Comment by JonathanGregory** [comment]
Dear @ethanrd 

Thanks for your comments. I've implemented your two small corrections. I agree with your larger suggestion. I have added a further paragraph to the introduction [NB the links do not work here in the issue!]

> This document is long, but it's not all immediately relevant when starting a proposal for changing the conventions, so please don't feel obliged to read it all at the outset. The most relevant parts are about the categories of changes in the [first section](#rules), the opening paragraphs of the [section on enhancements](#enhancement_rules), its subsections on the [way the discussion is run](#conduct) and [how it reaches a conclusion](#conclusions), and the [section on correcting defects](#defect_rules).

Is that suitable?

Jonathan

**Comment by JonathanGregory** [comment]
@ethanrd. Thanks also for telling me about the links. That's much better than linking the blob. Moreover, it seems that you can link a fragment of the page, which is even better e.g. the [present version of `CONTRIBUTING`](https://github.com/cf-convention/cf-conventions?tab=contributing-ov-file#contributing-to-the-netcdf-cf-conventions) and the [code of conduct](https://github.com/cf-convention/cf-conventions?tab=coc-ov-file#contributor-code-of-conduct).

**Comment by ethanrd** [comment]
Hi @JonathanGregory - Perfect. That looks good. One possible addition. It might be good to mention the issue labels which are described right after the change categories. Maybe just adding a bit to the second sentence:
> The most relevant parts are about the categories of changes in the [first section](#rules) **and the associated labels when creating an issue**, the opening paragraphs of the [section on enhancements](#enhancement_rules), ...

**Comment by JonathanGregory** [comment]
Thanks, @ethanrd, good idea. To make it easier to refer to, I have made the categories and labels text into its own subsection, and changed the sentence accordingly to:

> The most relevant parts are the start of the [first section](#rules), the [section on categories and labels for changes](#categories_and_labels), the opening paragraphs of the [section on enhancements](#enhancement_rules), its subsections on the [way the discussion is run](#conduct) and [how it reaches a conclusion](#conclusions), and the [section on correcting defects](#defect_rules).

See [the rendered modified version](https://github.com/JonathanGregory/cf-conventions/tree/issue369?tab=contributing-ov-file#contributing-to-the-cf-conventions), using the same magic.

**Comment by ethanrd** [comment]
Looks good! Thanks @JonathanGregory

**Comment by JonathanGregory** [comment]
Since there have been no substantial concerns raised about the proposed new consolidated `CONTRIBUTING` and rules document for the last three weeks, and enough support has been expressed for them, I believe we have agreed. Thank you to everyone. I will merge the new rules now. That will close this issue, and also lead to the conclusion of a couple of other issues in this repo, as listed [above](https://github.com/cf-convention/cf-conventions/issues/369#issuecomment-4236955388) by @erget. We can also proceed with the linked [website issue 632](https://github.com/cf-convention/cf-convention.github.io/issues/632).
